
From: Nick Piggin <nickpiggin@yahoo.com.au>

Shortcut escape from this queue if a hot task is met.  I'm skeptical that
it is good, but again it gives ppl something to test.


Index: linux-2.6.5/kernel/sched.c
===================================================================
--- linux-2.6.5.orig/kernel/sched.c	2004-04-16 15:06:46.000000000 +0200
+++ linux-2.6.5/kernel/sched.c	2004-04-16 15:06:46.000000000 +0200
@@ -1301,7 +1301,7 @@
 	if (idle == NEWLY_IDLE ||
 			sd->nr_balance_failed < sd->cache_nice_tries) {
 		if (task_hot(p, rq->timestamp_last_tick, sd))
-			return 0;
+			return -1;
 	}
 
 	return 1;
@@ -1320,7 +1320,7 @@
 {
 	prio_array_t *array, *dst_array;
 	struct list_head *head, *curr;
-	int idx, pulled = 0;
+	int ret, idx, pulled = 0;
 	task_t *tmp;
 
 	if (max_nr_move <= 0 || busiest->nr_running <= 1)
@@ -1359,7 +1359,12 @@
 
 	curr = curr->next;
 
-	if (!can_migrate_task(tmp, busiest, this_cpu, sd, idle)) {
+	ret = can_migrate_task(tmp, busiest, this_cpu, sd, idle);
+	if (ret == -1) {
+		idx++;
+		goto skip_bitmap;
+	}
+	if (!ret) {
 		if (curr != head)
 			goto skip_queue;
 		idx++;
